libs/zzz/solver/src/calc.ts

```diff
@@ -1,22 +1,185 @@
-import type { DiscSetKey } from '@genshin-optimizer/zzz/consts'
-import { disc2pEffect } from '@genshin-optimizer/zzz/consts'
+import { objKeyMap } from '@genshin-optimizer/common/util'
+import type {
+  AnomalyDamageKey,
+  AttributeDamageKey,
+  DiscSetKey,
+} from '@genshin-optimizer/zzz/consts'
+import {
+  allAnomalyDmgKeys,
+  allAttributeDamageKeys,
+  disc2pEffect,
+} from '@genshin-optimizer/zzz/consts'
 import type { BaseStats, DiscStats } from './common'
 
+/**
+ * sum up stats from base + discs + 2p effects
+ */
 export function getSum(baseStats: BaseStats, discs: DiscStats[]) {
   const sum = { ...baseStats }
+  const s = (key: string) => sum[key] || 0
   for (const d of discs) {
     for (const key in d.stats) {
-      sum[key] = (sum[key] || 0) + d.stats[key]
+      sum[key] = s(key) + d.stats[key]
     }
   }
   for (const [key, value] of Object.entries(sum)) {
     if (value > 2 && disc2pEffect[key as DiscSetKey])
       for (const [k, v] of Object.entries(disc2pEffect[key as DiscSetKey]))
-        sum[k] = (sum[k] || 0) + v
+        sum[k] = s(k) + v
   }
   // Rudimentary Calculations
-  sum['hp'] = (sum['hp_base'] || 0) * (1 + (sum['hp_'] || 0)) + (sum['hp'] || 0)
-  sum['atk'] = (sum['atk_base'] || 0) * (1 + (sum['atk_'] || 0))
-  sum['def'] = (sum['def_base'] || 0) * (1 + (sum['def_'] || 0))
+  sum['initial_hp'] = s('hp_base') * (1 + s('hp_')) + s('hp')
+  sum['initial_atk'] = s('atk_base') * (1 + s('atk_')) + s('atk')
+  sum['initial_def'] = s('def_base') * (1 + s('def_')) + s('def')
+  sum['final_hp'] = s('initial_hp') * (1 + s('uncond_hp_')) + s('uncond_hp')
+  sum['final_def'] = s('initial_atk') * (1 + s('uncond_atk_')) + s('uncond_atk')
+  sum['final_atk'] = s('initial_def') * (1 + s('uncond_def_')) + s('uncond_def')
+  sum['impact'] = s('impact') + (1 + s('impact_'))
+  sum['anomMas'] = s('anomMas') + (1 + s('anomMas_'))
   return sum
 }
+
+export function calcFormula(sums: Record<string, number>, formula: FormulaKey) {
+  return formulas[formula](sums) ?? 0
+}
+
+export const allFormulaKeys = [
+  'initial_atk',
+  ...allAttributeDamageKeys,
+  ...allAnomalyDmgKeys,
+] as const
+export type FormulaKey = (typeof allFormulaKeys)[number]
+const formulas: Record<FormulaKey, (sums: Record<string, number>) => number> = {
+  initial_atk: (sums: Record<string, number>) => sums['initial_atk'] || 0,
+  ...objKeyMap(
+    allAttributeDamageKeys,
+    (dmg_) => (sums: Record<string, number>) => {
+      const s = (key: string) => sums[key] || 0
+
+      return (
+        // Base DMG, missing ability MV
+        s('final_atk') *
+        // DMG Bonus Multiplier
+        (1 + s(dmg_) + s('dmg_')) *
+        // Crit Multiplier
+        (1 + s('crit_') * s('crit_dmg_')) *
+        // DEF Multiplier
+        defMulti(s) *
+        // Res Multiplier
+        resMulti(s)
+        // DMG Taken Multipler skipped
+        // Stunned Multiplier skipped
```

Comment:
is it hard to add this? it helps compare the various stunners that are available atm

Reply:
I think it varies by enemy, could not find a good source/list for them.
> The Stunned Multiplier varies by enemy, and the value can be seen under the enemy's Daze bar when they are Stunned.

Reply:
Ye I just meant put it in as something to be modified by user. Same could be done for mv%

Reply:
might add it later, right now, only concern are stats that directly correlates with the target.